File: packages/java/src/test/java/io/idrto/def/DefTest.java
package io.idrto.def;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.junit.jupiter.api.Test;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Locale;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class DefTest {
    private static JsonObject loadVectors() throws Exception {
        Path path = Path.of("..", "..", "vectors", "test-vectors.json").normalize();
        String json = Files.readString(path);
        return new Gson().fromJson(json, JsonObject.class);
    }

    private static Def.ErrorCode reasonToCode(String reason) {
        return switch (reason) {
            case "label_too_long" -> Def.ErrorCode.LABEL_TOO_LONG;
            case "invalid_escape" -> Def.ErrorCode.INVALID_ESCAPE;
            case "invalid_utf8" -> Def.ErrorCode.INVALID_UTF8;
            case "invalid_encoding" -> Def.ErrorCode.INVALID_ENCODING;
            case "not_decodable" -> Def.ErrorCode.NOT_DECODABLE;
            default -> throw new IllegalArgumentException("unknown reason: " + reason);
        };
    }

    @Test
    void encodeComponentVectors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("encode_component");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(c.get("encoded").getAsString(), Def.encodeComponent(c.get("input").getAsString()));
        }
    }

    @Test
    void decodeComponentVectors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("decode_component");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(c.get("decoded").getAsString(), Def.decodeComponent(c.get("input").getAsString()));
        }
    }

    @Test
    void encodeBodyVectors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("encode_body");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(c.get("encoded").getAsString(), Def.encodeBody(c.get("input").getAsString()));
        }
    }

    @Test
    void deterministicRoundTripsPreserveSeparators() throws Exception {
        for (int i = 0; i < 256; i++) {
            String component = "A" + i + "-x@用户" + (char) ('a' + (i % 26));
            String encodedComponent = Def.encodeComponent(component);
            assertFalse(encodedComponent.contains(Def.STRUCTURAL_SEPARATOR));
            assertEquals(component.toLowerCase(Locale.ROOT), Def.decodeComponent(encodedComponent));

            String input = component
                    + Def.STRUCTURAL_SEPARATOR.repeat(i % 4)
                    + (i % 3 == 0 ? "-" : "")
                    + "B" + (255 - i);
            String encodedBody = Def.encodeBody(input);
            assertEquals(separatorCount(input), separatorCount(encodedBody));
            assertEquals(input.toLowerCase(Locale.ROOT), Def.decodeBody(encodedBody));
        }

        Def.DefException ex = assertThrows(
                Def.DefException.class,
                () -> Def.decodeComponent("a--b")
        );
        assertEquals(Def.ErrorCode.INVALID_ESCAPE, ex.getCode());
    }

    private static int separatorCount(String value) {
        int count = 0;
        int offset = 0;
        int found;
        while ((found = value.indexOf(Def.STRUCTURAL_SEPARATOR, offset)) >= 0) {
            count++;
            offset = found + Def.STRUCTURAL_SEPARATOR.length();
        }
        return count;
    }

    @Test
    void decodeBodyVectors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("decode_body");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(c.get("decoded").getAsString(), Def.decodeBody(c.get("input").getAsString()));
        }
    }

    @Test
    void decodeBodyErrors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("decode_body_errors");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            Def.DefException ex = assertThrows(
                    Def.DefException.class,
                    () -> Def.decodeBody(c.get("input").getAsString())
            );
            assertEquals(reasonToCode(c.get("reason").getAsString()), ex.getCode());
        }
    }

    @Test
    void encodeProfileVectors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("encode_profile");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(
                    c.get("encoded").getAsString(),
                    Def.encodeProfile(c.get("input").getAsString(), Def.IDRTO_HASH_MARKER)
            );
        }
    }

    @Test
    void encodeProfileHashVectors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("encode_profile_hash");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(
                    c.get("encoded").getAsString(),
                    Def.encodeProfile(c.get("input").getAsString(), Def.IDRTO_HASH_MARKER)
            );
        }
    }

    @Test
    void encodeProfileErrors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("encode_profile_errors");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            Def.DefException ex = assertThrows(
                    Def.DefException.class,
                    () -> Def.encodeProfile(c.get("input").getAsString(), Def.IDRTO_HASH_MARKER)
            );
            assertEquals(reasonToCode(c.get("reason").getAsString()), ex.getCode());
        }
    }

    @Test
    void decodeProfileVectors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("decode_profile");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(
                    c.get("decoded").getAsString(),
                    Def.decodeProfile(c.get("input").getAsString(), Def.IDRTO_HASH_MARKER)
            );
        }
    }

    @Test
    void decodeProfileErrors() throws Exception {
        JsonArray cases = loadVectors().getAsJsonArray("decode_profile_errors");
        for (var element : cases) {
            JsonObject c = element.getAsJsonObject();
            Def.DefException ex = assertThrows(
                    Def.DefException.class,
                    () -> Def.decodeProfile(c.get("input").getAsString(), Def.IDRTO_HASH_MARKER)
            );
            assertEquals(reasonToCode(c.get("reason").getAsString()), ex.getCode());
        }
    }

    @Test
    void profileIsGenericAndHashesMarkerCollisions() throws Exception {
        assertEquals("xn--name", Def.encodeProfile("XN--Name"));
        assertEquals("xn--name", Def.decodeProfile("xn--name"));

        String marker = "abc--";
        String collision = Def.encodeProfile("abc--value", marker);
        assertTrue(collision.startsWith(marker));
        Def.DefException ex = assertThrows(
                Def.DefException.class,
                () -> Def.decodeProfile(collision, marker)
        );
        assertEquals(Def.ErrorCode.NOT_DECODABLE, ex.getCode());
    }

    @Test
    void idrtoShortcuts() throws Exception {
        JsonObject vectors = loadVectors();

        for (var element : vectors.getAsJsonArray("encode_profile")) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(c.get("encoded").getAsString(), Def.encode(c.get("input").getAsString()));
        }

        for (var element : vectors.getAsJsonArray("decode_profile")) {
            JsonObject c = element.getAsJsonObject();
            assertEquals(c.get("decoded").getAsString(), Def.decode(c.get("input").getAsString()));
        }
    }
}

File: packages/go/def/def_test.go
package def_test

import (
	"errors"
	"testing"

	"github.com/idrto/dns-encoded-format/go/def"
)

func TestVectors(t *testing.T) {
	v, err := def.LoadVectorsForTest()
	if err != nil {
		t.Fatal(err)
	}

	for _, c := range v.EncodeComponent {
		got := def.EncodeComponent(c.Input)
		if got != c.Encoded {
			t.Fatalf("EncodeComponent(%q) = %q, want %q", c.Input, got, c.Encoded)
		}
	}

	for _, c := range v.DecodeComponent {
		got, err := def.DecodeComponent(c.Input)
		if err != nil {
			t.Fatalf("DecodeComponent(%q): %v", c.Input, err)
		}
		if got != c.Decoded {
			t.Fatalf("DecodeComponent(%q) = %q, want %q", c.Input, got, c.Decoded)
		}
	}

	for _, c := range v.EncodeBody {
		got := def.EncodeBody(c.Input)
		if got != c.Encoded {
			t.Fatalf("EncodeBody(%q) = %q, want %q", c.Input, got, c.Encoded)
		}
	}

	for _, c := range v.DecodeBody {
		got, err := def.DecodeBody(c.Input)
		if err != nil {
			t.Fatalf("DecodeBody(%q): %v", c.Input, err)
		}
		if got != c.Decoded {
			t.Fatalf("DecodeBody(%q) = %q, want %q", c.Input, got, c.Decoded)
		}
	}

	for _, c := range v.DecodeBodyErrors {
		_, err := def.DecodeBody(c.Input)
		if !errors.Is(err, reasonToError(c.Reason)) {
			t.Fatalf("DecodeBody(%q): got %v, want %v", c.Input, err, c.Reason)
		}
	}

	for _, c := range v.EncodeProfile {
		got, err := def.EncodeProfile(c.Input, def.IdrtoHashMarker)
		if err != nil {
			t.Fatalf("EncodeProfile(%q): %v", c.Input, err)
		}
		if got != c.Encoded {
			t.Fatalf("EncodeProfile(%q) = %q, want %q", c.Input, got, c.Encoded)
		}
	}

	for _, c := range v.EncodeProfileHash {
		got, err := def.EncodeProfile(c.Input, def.IdrtoHashMarker)
		if err != nil {
			t.Fatalf("EncodeProfile(%q): %v", c.Input, err)
		}
		if got != c.Encoded {
			t.Fatalf("EncodeProfile(%q) = %q, want %q", c.Input, got, c.Encoded)
		}
	}

	for _, c := range v.EncodeProfileErrors {
		_, err := def.EncodeProfile(c.Input, def.IdrtoHashMarker)
		if !errors.Is(err, reasonToError(c.Reason)) {
			t.Fatalf("EncodeProfile(%q): got %v, want %v", c.Input, err, c.Reason)
		}
	}

	for _, c := range v.DecodeProfile {
		got, err := def.DecodeProfile(c.Input, def.IdrtoHashMarker)
		if err != nil {
			t.Fatalf("DecodeProfile(%q): %v", c.Input, err)
		}
		if got != c.Decoded {
			t.Fatalf("DecodeProfile(%q) = %q, want %q", c.Input, got, c.Decoded)
		}
	}

	for _, c := range v.DecodeProfileErrors {
		_, err := def.DecodeProfile(c.Input, def.IdrtoHashMarker)
		if !errors.Is(err, reasonToError(c.Reason)) {
			t.Fatalf("DecodeProfile(%q): got %v, want %v", c.Input, err, c.Reason)
		}
	}
}

func TestDeterministicCoreProperties(t *testing.T) {
	alphabet := []string{"", "a", "B", "-", "--", "@", "é", "用户", "😊"}
	values := append([]string(nil), alphabet...)
	for _, left := range alphabet {
		for _, right := range alphabet {
			values = append(values, left+right, left+"--"+right)
		}
	}

	for _, value := range values {
		encoded := def.EncodeBody(value)
		decoded, err := def.DecodeBody(encoded)
		if err != nil {
			t.Fatalf("DecodeBody(EncodeBody(%q)): %v", value, err)
		}
		if decoded != asciiLower(value) {
			t.Fatalf("DecodeBody(EncodeBody(%q)) = %q", value, decoded)
		}
		if countSeparators(encoded) != countSeparators(value) {
			t.Fatalf("separator count changed for %q: %q", value, encoded)
		}
		if countSeparators(def.EncodeComponent(value)) != 0 {
			t.Fatalf("EncodeComponent(%q) emitted a raw separator", value)
		}
	}

	if _, err := def.DecodeComponent(def.StructuralSeparator); !errors.Is(err, def.ErrInvalidEscape) {
		t.Fatalf("DecodeComponent(%q): got %v", def.StructuralSeparator, err)
	}
}

func TestGenericProfileContract(t *testing.T) {
	encoded, err := def.Encode("xn--value")
	if err != nil || encoded != "xn--value" {
		t.Fatalf("Encode(%q) = %q, %v", "xn--value", encoded, err)
	}
	decoded, err := def.Decode(encoded)
	if err != nil || decoded != "xn--value" {
		t.Fatalf("Decode(%q) = %q, %v", encoded, decoded, err)
	}

	marker := "abc--"
	collision, err := def.EncodeProfile("abc--value", marker)
	if err != nil {
		t.Fatalf("Encode marker collision: %v", err)
	}
	if len(collision) != len(marker)+def.HashBodyLength {
		t.Fatalf("marker collision did not hash: %q", collision)
	}
	if _, err := def.DecodeProfile(collision, marker); !errors.Is(err, def.ErrNotDecodable) {
		t.Fatalf("Decode marker collision: got %v", err)
	}
}

func asciiLower(value string) string {
	bytes := []byte(value)
	for i, b := range bytes {
		if b >= 'A' && b <= 'Z' {
			bytes[i] = b + ('a' - 'A')
		}
	}
	return string(bytes)
}

func countSeparators(value string) int {
	count := 0
	for i := 0; i+1 < len(value); {
		if value[i:i+2] == def.StructuralSeparator {
			count++
			i += 2
		} else {
			i++
		}
	}
	return count
}

func reasonToError(reason string) error {
	switch reason {
	case "label_too_long":
		return def.ErrLabelTooLong
	case "invalid_escape":
		return def.ErrInvalidEscape
	case "invalid_utf8":
		return def.ErrInvalidUTF8
	case "invalid_encoding":
		return def.ErrInvalidEncoding
	case "not_decodable":
		return def.ErrNotDecodable
	default:
		panic("unknown reason: " + reason)
	}
}
